src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -43,16 +43,40 @@ protected override VSInternalDiagnosticReport CreateReport(TextDocumentIdentifie
                 Supersedes = WorkspaceDiagnosticIdentifier,
             };
 
-        protected override VSInternalDiagnosticParams[]? GetPreviousResults(VSInternalDocumentDiagnosticsParams diagnosticsParams)
-            => new[] { diagnosticsParams };
+        protected override PreviousResult[]? GetPreviousResults(VSInternalDocumentDiagnosticsParams diagnosticsParams)
+        {
+            if (diagnosticsParams.PreviousResultId != null && diagnosticsParams.TextDocument != null)
+            {
+                return new[] { new PreviousResult(diagnosticsParams.PreviousResultId, diagnosticsParams.TextDocument) };
+            }
 
-        protected override IProgress<VSInternalDiagnosticReport[]>? GetProgress(VSInternalDocumentDiagnosticsParams diagnosticsParams)
-            => diagnosticsParams.PartialResultToken;
+            return null;
+        }
 
         protected override DiagnosticTag[] ConvertTags(DiagnosticData diagnosticData)
             => ConvertTags(diagnosticData, potentialDuplicate: false);
 
         protected override ImmutableArray<Document> GetOrderedDocuments(RequestContext context)
+        {
+            return GetRequestedDocument(context);
+        }
+
+        protected override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+            RequestContext context, Document document, Option2<DiagnosticMode> diagnosticMode, CancellationToken cancellationToken)
+        {
+            // For open documents, directly use the IDiagnosticAnalyzerService.  This will use the actual snapshots
+            // we're passing in.  If information is already cached for that snapshot, it will be returned.  Otherwise,
+            // it will be computed on demand.  Because it is always accurate as per this snapshot, all spans are correct
+            // and do not need to be adjusted.
+            return _analyzerService.GetDiagnosticsForSpanAsync(document, range: null, cancellationToken: cancellationToken);
```

Comment:
note for future LSP meetings.  It's a shame we can't benefit from the range bit.  it woudl be nice if the client only cared about what was in view and scoped down to that.